libvast/builtins/commands/rebuild.cpp

```diff
@@ -56,178 +56,41 @@ class rebatch_operator final : public pipeline_operator {
   /// @param schema The schema of the input and output table slices.
   /// @param desired_batch_size The desired output batch size; guaranteed to be
   /// exactly matched for all but the last output batch.
-  rebatch_operator(type schema, int64_t desired_batch_size)
-    : schema_{std::move(schema)},
-      desired_batch_size_{desired_batch_size},
-      arrow_schema_{schema_.to_arrow_schema()},
-      builder_{caf::get<record_type>(schema_).make_arrow_builder(
-        arrow::default_memory_pool())} {
-    const auto resize_result = builder_->Resize(desired_batch_size_);
-    VAST_ASSERT(resize_result.ok(), resize_result.ToString().c_str());
+  rebatch_operator(type schema, size_t desired_batch_size)
+    : schema_{std::move(schema)}, desired_batch_size_{desired_batch_size} {
+    // nop
   }
 
   caf::error
   add(type schema, std::shared_ptr<arrow::RecordBatch> batch) override {
-    VAST_ASSERT(schema == schema_);
-    VAST_ASSERT(builder_ != nullptr);
-    const auto batch_size = batch->num_rows();
-    if (builder_->length() == 0 && batch_size == desired_batch_size_) {
-      result_.push_back(batch);
-      return caf::none;
-    }
-    for (auto offset = int64_t{}; offset < batch->num_rows();) {
-      const auto length = std::min(desired_batch_size_ - builder_->length(),
-                                   batch_size - offset);
-      append_slice_to_builder(batch, offset, length);
-      offset += length;
-      VAST_ASSERT(builder_->length() <= desired_batch_size_);
-      if (builder_->length() == desired_batch_size_)
-        finish_builder();
+    auto slice = table_slice{batch, std::move(schema)};
+    const auto buffered_rows = rows(buffer_) + slice.rows();
+    if (buffered_rows < desired_batch_size_) {
+      buffer_.push_back(std::move(slice));
+    } else {
+      const auto remainder = desired_batch_size_ - buffered_rows;
+      auto [head, tail] = split(slice, slice.rows() - remainder);
```

Comment:
This split should be purely logical (e.g., based on bitmaps or tracking row offsets). I would not expect this function to perform expensive work other than adding data to a builder.

Reply:
Yeah, `split` basically just calls `batch->Slice(...)` twice now, so this has O(1). In most cases we should be in the `if` case instead of the `else` here, as after a single rebuild data should have the desired batch size already.

NB: I didn't change the rebuild algorithm at all, it now just calls the functions on table slices rather than doing everything on the underlying record batches, as that now maps to the same set of operations.